### PR TITLE
4343 put error in state

### PR DESCRIPTION
**Pull request by brett-varney**
**High level description:**

Testing revealed error condition not reacting properly, this resolves that

**Technical details:**

put isError in state instead of local

To test, change api request locally to:
```
export const fetchObligationsByAwardType = (code, fy) => {
    return {
        promise: Promise.resolve({
            data: {
                "total_aggregated_amount": 24334157902.1,
                "results": [
                    { "category": "contracts", "aggregated_amount": 3219784714.6 },
                    { "category": "direct_payments", "aggregated_amount": 555685440.08 },
                    { "category": "grants", "aggregated_amount": 20172163677.11 },
                    { "category": "idv", "aggregated_amount": 387656184.37 },
                    { "category": "loans", "aggregated_amount": 0.0 },
                    { "category": "other", "aggregated_amount": 0 }
                ]
            }
        })
    };
}
```

**JIRA Ticket:**
[DEV-4343](https://federal-spending-transparency.atlassian.net/browse/DEV-4343)

**Mockup:**
in ticket

The following are ALL required for the PR to be merged:

Author:
- [x] Linked to this PR in JIRA ticket
- [pre-done] Scheduled demo including Design/Testing/Front-end OR Provided instructions for testing in JIRA and PR `if applicable`
- [x] Verified cross-browser compatibility: Chrome, Safari, Firefox, Internet Explorer, Edge
- [x] Verified mobile/tablet/desktop/monitor responsiveness
- [x] Verified that this PR does not create any *new* accessibility issues (via Axe Chrome extension)
- [x] Added Unit Tests for helper functions, reducers, models and Container/Component Interactivity Expectations `if applicable` [React Testing Library](react-testing-library.md)
- [na] [API contract](https://github.com/fedspendingtransparency/usaspending-api/tree/dev/usaspending_api/api_contracts) updated `if applicable`
- [na] [Component Library Integration Status Report](https://github.com/fedspendingtransparency/data-act-documentation/blob/data-transparency-ui/frontend_apps/component-library-integration-status.md) updated `if applicable`

Reviewer(s):
- [na] Design review complete `if applicable`
- [na] [API #1234](https://github.com/fedspendingtransparency/usaspending-api/pull/1234) merged concurrently `if applicable`
- [x] Code review complete

### CODE REVIEW

**Review by ebdabbs — CHANGES_REQUESTED** [review]
I think we'll also need to reset the error state to false after a successful request. With React dev tools (in Chrome) if I change the error state to `true` and then update the fiscal year filter to kick off a new request, the error message still displays.

**Comment by brett-varney** [comment]
> 
> 
> I think we'll also need to reset the error state to false after a successful request. With React dev tools (in Chrome) if I change the error state to `true` and then update the fiscal year filter to kick off a new request, the error message still displays.

Thanks, done

File: src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx [hunk 1]
```diff
@@ -22,10 +22,11 @@ const propTypes = {
 
 export default function ObligationsByAwardTypeContainer({ fiscalYear, windowWidth }) {
     const [isLoading, setLoading] = React.useState(true);
+    const [isError, setError] = React.useState(false);
     const { toptierCode } = useSelector((state) => state.agencyV2.overview);
-    let isError = false;
 
     if (toptierCode) {
+        let error = false;
```
Comment (jonhill13) [inline src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx:29]: This should set the error. We have access to whether or not there is an error in `isError`. 
`setError(false)`

File: src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx [hunk 2]
```diff
@@ -95,11 +96,12 @@ export default function ObligationsByAwardTypeContainer({ fiscalYear, windowWidt
                         break;
                     default:
                         console.error(`Category name from API not recognized: ${d.category}`);
-                        isError = true;
+                        error = true;
```
Comment (jonhill13) [inline src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx:99]: This should set the error.
`setError(false)`

File: src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx (hunk 1 above)
Comment (jonhill13) [reply src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx:29]: We should also watch the loading here, if the fiscal year is updated we might need to set loading to be true too.

File: src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx [hunk 3]
```diff
@@ -22,10 +22,11 @@ const propTypes = {
 
 export default function ObligationsByAwardTypeContainer({ fiscalYear, windowWidth }) {
     const [isLoading, setLoading] = React.useState(true);
+    const [isError, setError] = React.useState(false);
     const { toptierCode } = useSelector((state) => state.agencyV2.overview);
-    let isError = false;
 
     if (toptierCode) {
+        let error = false;
         fetchObligationsByAwardType(toptierCode, fiscalYear).promise.then((res) => {
```
Comment (jonhill13) [inline src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx:30]: We should be catching the error from the api with a `.catch()`.

**Review by jonhill13 — CHANGES_REQUESTED** [review]

File: src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx (hunk 1 above)
Comment (brett-varney) [reply src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx:29]: I agree, but that causes an infinite rendering loop

Comment (jonhill13) [reply src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx:29]: We can update the component with some additional state, refs, effects and catching api errors to look something like this.

```
export default function ObligationsByAwardTypeContainer({ fiscalYear, windowWidth }) {
    const [categoriesForGraph, setCategoriesForGraph] = useState([]);
    const [detailsForGraph, setDetailsForGraph] = useState([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(false);
    const obligationsByAwardTypeRequest = useRef(null);
    const { toptierCode } = useSelector((state) => state.agencyV2.overview);

    useEffect(() => () => {
        if (obligationsByAwardTypeRequest.current) {
            obligationsByAwardTypeRequest.current.cancel();
        }
    }, []);

    const getObligationsByAwardType = () => {
        if (obligationsByAwardTypeRequest.current) {
            obligationsByAwardTypeRequest.current.cancel();
        }
        if (error) setError(false);
        if (!loading) setLoading(true);
        obligationsByAwardTypeRequest.current = fetchObligationsByAwardType(toptierCode, fiscalYear);
        obligationsByAwardTypeRequest.current.promise.then((res) => {
            const categories = defaultCategories;
            const details = defaultDetails;
            res.data.results.forEach((d) => {
                switch (d.category) {
                    case 'grants':
                        categories[0].value += d.aggregated_amount;
                        details[0].value = d.aggregated_amount;
                        break;
                    case 'loans':
                        categories[0].value += d.aggregated_amount;
                        details[1].value = d.aggregated_amount;
                        break;
                    case 'direct_payments':
                        categories[0].value += d.aggregated_amount;
                        details[2].value = d.aggregated_amount;
                        break;
                    case 'other':
                        categories[0].value += d.aggregated_amount;
                        details[3].value = d.aggregated_amount;
                        break;
                    case 'idvs':
                        categories[1].value += d.aggregated_amount;
                        details[4].value = d.aggregated_amount;
                        break;
                    case 'contracts':
                        categories[1].value += d.aggregated_amount;
                        details[5].value = d.aggregated_amount;
                        break;
                    default:
                        console.error(`Category name from API not recognized: ${d.category}`);
                        setError(true);
                }
            });
            setCategoriesForGraph(categories);
            setDetailsForGraph(details);
            setLoading(false);
            obligationsByAwardTypeRequest.current = null;
        }).catch((e) => {
            if (!isCancel(e)) {
                setError(true);
                setLoading(false);
                obligationsByAwardTypeRequest.current = null;
                console.error(e);
            }
        });
    };

    useEffect(() => {
        if (toptierCode) {
            getObligationsByAwardType();
        }
    }, [fiscalYear, toptierCode]);


    return (<>
        { loading && <LoadingMessage />}
        { error && <ErrorMessage />}
        { !loading && !error &&
            <ObligationsByAwardType
                outer={categoriesForGraph}
                inner={detailsForGraph}
                windowWidth={windowWidth} />
        }
    </>);
}
```

Updating the categories and details to state should allow react to watch for those updates and rerender the children.

Referencing the request gives us control over it allowing us to cancel any current requests when unmounting.

Moving the request logic to its own function allows us to use an effect to make an api call only when certain data changes.

Catching the response error allows us to ignore and cancel requests we may make and updates the user when there is an api error.

**Review by jonhill13 — CHANGES_REQUESTED** [review]

File: src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx (hunk 1 above)
Comment (brett-varney) [reply src/js/containers/agencyV2/visualizations/ObligationsByAwardTypeContainer.jsx]: Even better ideas; done

**Review by jonhill13 — DISMISSED** [review]

**Review by ebdabbs — CHANGES_REQUESTED** [review]
These changes look good, but we should have unit tests for this file

**Review by ebdabbs — APPROVED** [review]

**Pull request merged**